modules/Graphs.tla

```diff
@@ -64,6 +64,185 @@ IsTreeWithRoot(G, r) ==
+
+(**
+ * Returns the set of nodes that are immediate successors of any node in S.
+ *
+ * @param G: A directed graph as a record.
+ * @param S: A set of nodes.
+ * @return: The union of successors of all nodes in S.
+ *
+ * Example:
+ *   G = [node |-> {1, 2, 3}, edge |-> {<<1, 2>>, <<1, 3>>}]
+ *   AllSuccessors(G, {1, 2}) = {2, 3}
+ *)
+AllSuccessors(G, S) == UNION {Successors(G, n): n \in S}
+
+(**
+ * Returns the set of nodes that are immediate predecessors of node n in G.
+ *
+ * @param G: A directed graph as a record.
+ * @param n: A node.
+ * @return: The set of nodes with edges pointing into n.
+ *
+ * Example:
+ *   G = [node |-> {1, 2, 3}, edge |-> {<<2, 1>>, <<3, 1>>}]
+ *   Predecessors(G, 1) = {2, 3}
+ *)
+Predecessors(G, n) == {m \in G.node: << m, n >> \in G.edge}
+
+(**
+ * Returns the set of nodes that are immediate predecessors of any node in S.
+ *
+ * @param G: A directed graph as a record.
+ * @param S: A set of nodes.
+ * @return: The union of predecessors of all nodes in S.
+ *
+ * Example:
+ *   G = [node |-> {1, 2, 3}, edge |-> {<<2, 1>>, <<3, 1>>}]
+ *   AllPredecessors(G, {1, 2}) = {2, 3}
+ *)
+AllPredecessors(G, S) == UNION {Predecessors(G, n): n \in S}
+
+(**
+ * Returns the in-degree of node n in directed graph G.
+ *
+ * @param G: A directed graph as a record.
+ * @param n: A node.
+ * @return: The number of incoming edges to n.
+ *
+ * Example:
+ *   G = [node |-> {1, 2, 3}, edge |-> {<<2, 1>>, <<3, 1>>}]
+ *   InDegree(G, 1) = 2
+ *)
+InDegree(G, n) == Cardinality(Predecessors(G, n))
+
+(**
+ * Returns the out-degree of node n in directed graph G.
+ *
+ * @param G: A directed graph as a record.
+ * @param n: A node.
+ * @return: The number of outgoing edges from n.
+ *
+ * Example:
+ *   G = [node |-> {1, 2, 3}, edge |-> {<<1, 2>>, <<1, 3>>}]
+ *   OutDegree(G, 1) = 2
+ *)
+OutDegree(G, n) == Cardinality(Successors(G, n))
+
+(**
+ * Returns the set of root nodes of G.
+ *
+ * @param G: A directed graph as a record.
+ * @return: The set of nodes with no incoming edges.
+ *)
+Roots(G) == {n \in G.node: Predecessors(G, n) = {}}
+
+(**
+ * Returns the set of leaf nodes of G.
+ *
+ * @param G: A directed graph as a record.
+ * @return: The set of nodes with no outgoing edges.
+ *)
+Leaves(G) == {n \in G.node: Successors(G, n) = {}}
+-----------------------------------------------------------------------------
+(**
+ * The graph with no nodes and no edges.
+ *
+ * @return: The empty graph as a record.
+ *)
+EmptyGraph == [node |-> {}, edge |-> {}]
+
+(**
+ * The set of all possible labeled directed graphs whose node set is S.
+ *
+ * @param S: A set of nodes.
+ * @return: The set of all labeled directed graphs over S.
+ *
+ * Example:
+ *   Graphs({1, 2}) = {
+ *     [node |-> {1, 2}, edge |-> {}],
+ *     [node |-> {1, 2}, edge |-> {<<1, 2>>}],
+ *     [node |-> {1, 2}, edge |-> {<<1, 2>>, <<2, 1>>}],
+ *     ...
+ *   }
+ *)
+Graphs(S) == [node: {S}, edge: SUBSET (S \X S)]

```

Comment:
This should certainly be ``[node: S, ....]``, otherwise the set of nodes is a singleton. Note that the test at line 11 of module GraphTests doesn't catch this because it holds trivially.

Reply:
```tla
IsDirectedGraph(G) == G = [node |-> G.node, edge |-> G.edge] /\ G.edge \subseteq (G.node \X G.node) IN \A g \in Graphs({1,2})

(tla+) LET Graphs(S) == [node: S, edge: SUBSET (S \X S)]  IN \A g \in Graphs({1,2}) : IsDirectedGraph(g)
Error evaluating expression: 'LET Graphs(S) == [node: S, edge: SUBSET (S \X S)] : IsDirectedGraph(g)'
Attempted to check if the value: 1 is an element of the integer 1

(tla+) LET Graphs(S) == [node: {S}, edge: SUBSET (S \X S)] IN \A g \in Graphs({1,2}) : IsDirectedGraph(g)
TRUE
```